### ISSUE TITLE
AccessControlException from Weld code when Java 2 security is enabled

### ISSUE THREAD

**Issue by andymc12** [OpenLiberty/open-liberty#690]
While doing some Java 2 security testing in the JAX-RS 2.X buckets, I ran into this failure - this occurs in the com.ibm.ws.jaxrs.2.0.cdi_fat bucket when Java 2 security is enabled and the JAX-RS 2.X servers are removed from the exempt list.

```
Stack Dump = java.security.AccessControlException: Access denied ("java.lang.RuntimePermission" "accessDeclaredMembers")
	at java.security.AccessController.throwACE(AccessController.java:157)
	at java.security.AccessController.checkPermissionHelper(AccessController.java:217)
	at java.security.AccessController.checkPermission(AccessController.java:349)
	at java.lang.SecurityManager.checkPermission(SecurityManager.java:562)
	at java.lang.Class.checkMemberAccess(Class.java:200)
	at java.lang.Class.getEnclosingMethod(Class.java:2912)
	at org.jboss.weld.util.reflection.Reflections.isStaticNestedClass(Reflections.java:402)
	at org.jboss.weld.util.reflection.Reflections.isTopLevelOrStaticNestedClass(Reflections.java:422)
	at org.jboss.weld.util.reflection.Reflections.isCacheable(Reflections.java:92)
	at org.jboss.weld.manager.BeanManagerImpl.getBeans(BeanManagerImpl.java:557)
	at com.ibm.ws.jaxrs20.cdi.fat.basic10.HelloWorldResource.getMessage2(HelloWorldResource.java:162)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:95)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
	at java.lang.reflect.Method.invoke(Method.java:508)
	at com.ibm.ws.jaxrs20.cdi.component.JaxRsFactoryImplicitBeanCDICustomizer.serviceInvoke(JaxRsFactoryImplicitBeanCDICustomizer.java:293)
	at com.ibm.ws.jaxrs20.server.LibertyJaxRsServerFactoryBean.performInvocation(LibertyJaxRsServerFactoryBean.java:618)
	at com.ibm.ws.jaxrs20.server.LibertyJaxRsInvoker.performInvocation(LibertyJaxRsInvoker.java:117)
	at org.apache.cxf.service.invoker.AbstractInvoker.invoke(AbstractInvoker.java:96)
	at com.ibm.ws.jaxrs20.server.LibertyJaxRsInvoker.invoke(LibertyJaxRsInvoker.java:251)
	at org.apache.cxf.jaxrs.JAXRSInvoker.invoke(JAXRSInvoker.java:189)
	at com.ibm.ws.jaxrs20.server.LibertyJaxRsInvoker.invoke(LibertyJaxRsInvoker.java:422)
	at org.apache.cxf.jaxrs.JAXRSInvoker.invoke(JAXRSInvoker.java:99)
	at org.apache.cxf.interceptor.ServiceInvokerInterceptor$1.run(ServiceInvokerInterceptor.java:61)
	at org.apache.cxf.interceptor.ServiceInvokerInterceptor.handleMessage(ServiceInvokerInterceptor.java:99)
	at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:308)
	at org.apache.cxf.transport.ChainInitiationObserver.onMessage(ChainInitiationObserver.java:124)
	at org.apache.cxf.transport.http.AbstractHTTPDestination.invoke(AbstractHTTPDestination.java:274)
	at com.ibm.ws.jaxrs20.endpoint.AbstractJaxRsWebEndpoint.invoke(AbstractJaxRsWebEndpoint.java:134)
	at com.ibm.websphere.jaxrs.server.IBMRestServlet.handleRequest(IBMRestServlet.java:146)
	at com.ibm.websphere.jaxrs.server.IBMRestServlet.doGet(IBMRestServlet.java:112)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
	at com.ibm.websphere.jaxrs.server.IBMRestServlet.service(IBMRestServlet.java:96)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.service(ServletWrapper.java:1255)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:743)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:440)
	at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1154)
	at com.ibm.ws.webcontainer.webapp.WebApp.handleRequest(WebApp.java:4962)
	at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.handleRequest(DynamicVirtualHost.java:314)
	at com.ibm.ws.webcontainer.WebContainer.handleRequest(WebContainer.java:995)
	at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.run(DynamicVirtualHost.java:279)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink$TaskWrapper.run(HttpDispatcherLink.java:957)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.wrapHandlerAndExecute(HttpDispatcherLink.java:357)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.ready(HttpDispatcherLink.java:316)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleDiscrimination(HttpInboundLink.java:499)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleNewRequest(HttpInboundLink.java:433)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.processRequest(HttpInboundLink.java:313)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.ready(HttpInboundLink.java:284)
	at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.sendToDiscriminators(NewConnectionInitialReadCallback.java:165)
	at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.complete(NewConnectionInitialReadCallback.java:74)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.requestComplete(WorkQueueManager.java:503)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.attemptIO(WorkQueueManager.java:573)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.workerRun(WorkQueueManager.java:928)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager$Worker.run(WorkQueueManager.java:1017)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1153)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.lang.Thread.run(Thread.java:785)
```

**Comment by benjamin-confino** [comment]
Weld are aware of the fact that they need to be more explicit about the security related rules. See:

https://issues.jboss.org/browse/CDI-486
https://issues.jboss.org/browse/CDI-699

My recommendation is to give the test app the right privileges until the above issues are resolved.

**Comment by Azquelt** [comment]
To summarize the contents of the issues Ben linked:

The CDI SPI (which includes BeanManager) allows you to access stuff which you shouldn't have access to. They should probably do something better here, but they can't just put `doPriv` around everything because this would allow apps to do reflection through the CDI SPI without the appropriate permissions.

In this particular example, you could call `getBeans()` to get a `Bean` and then call `getInjectionPoint().getAnnotated()` to get hold of the (possibly private) field or method that is an injection point of the bean.

You could also call `AnnotatedField.getDeclaringType().getMethods()` to get hold of all of the methods on that bean, which you shouldn't be able to do without the `accessDeclaredMembers` permission.

To summarize again, if you're using anything in `javax.enterprise.inject.spi`, you probably need to give your app permission to perform reflection.

**Comment by Azquelt** [comment]
@andymc12 sorry for the slow turnaround, but we came to the conclusion that, for the moment at least, you need to give your app the permission it's complaining about. If weld just added `doPriv` calls around the reflection calls that would allow you to bypass all of the checks that prevent your app doing reflection directly.

Possibly a future CDI update will address this and do its own security checking to allow you to access most of the SPI functions, with the exception of those that would allow you to bypass permissions checks.

**Comment by andymc12** [comment]
@Azquelt @benjamin-confino Thanks for that update.  Given the response from Weld, should the CDI 1.2 Javadocs be updated to indicate that the `BeanManager.getBeans(...)` method will throw an AccessControlException if a security manager is present?
